### PR TITLE
Formal & Power Tools

### PR DESCRIPTION
**Pull request by donggyukim**
Addresses #135. This is not ready to be merged, but for code review.

### CODE REVIEW

**Comment by edwardcwang** [comment]
Can you add (via pull request) an example script/config for using it in https://github.com/ucb-bar/hammer-nda-saed32-aspire-synthesis ? It will help me evaluate this PR :)

File: src/hammer-vlsi/cli_driver.py [hunk 1]
```diff
@@ -121,6 +129,23 @@ def create_par_action(self, custom_hooks: List[HammerToolHookAction],
...
+        hooks = self.get_extra_synthesis_hooks() + custom_hooks  # type: List[HammerToolHookAction]
+        return self.create_action("formal", hooks if len(hooks) > 0 else None,
+                                  pre_action_func, post_load_func, post_run_func)
+
+    def create_power_action(self, custom_hooks: List[HammerToolHookAction],
+                                pre_action_func: Optional[Callable[[HammerDriver], None]] = None,
+                                post_load_func: Optional[Callable[[HammerDriver], None]] = None,
+                                post_run_func: Optional[Callable[[HammerDriver], None]] = None
+                                ) -> CLIActionType:
+        hooks = self.get_extra_synthesis_hooks() + custom_hooks  # type: List[HammerToolHookAction]
```
Comment (colinschmidt) [inline src/hammer-vlsi/cli_driver.py]: These extra hooks should use `get_extra_power_hooks` method or the `get_extra_formal_hooks` right?

File: src/hammer-vlsi/cli_driver.py [hunk 2]
```diff
@@ -360,6 +400,11 @@ def args_to_driver(self, args: dict,
         if top_module is not None:
             config['synthesis.inputs.top_module'] = top_module
 
+        # Waveform files
+        waveforms = args['waveform']
+        if isinstance(waveforms, List) and len(waveforms) > 0:
+            config['power.inputs.waveform_files'] = waveforms
```
Comment (colinschmidt) [inline src/hammer-vlsi/cli_driver.py]: I think this should follow the `verilogs` example more closely. i.e. including the `list` conversion

File: src/hammer-vlsi/hammer_driver.py [hunk 3]
```diff
@@ -148,6 +153,80 @@ def update_tool_configs(self) -> None:
         tools = reduce(lambda a, b: a + b, list(self.tool_configs.values()))
         self.database.update_tools(tools)
 
+    def load_formal_tool(self, run_dir: str = "") -> bool:
+        """
+        Load the place and route tool based on the given database.
```
Comment (colinschmidt) [inline src/hammer-vlsi/hammer_driver.py]: `place and route` => `formal verification`

File: src/hammer-vlsi/hammer_driver.py [hunk 4]
```diff
@@ -148,6 +153,80 @@ def update_tool_configs(self) -> None:
...
+
+        :param run_dir: Directory to use for the tool run_dir. Defaults to the run_dir passed in the HammerDriver
+                        constructor.
+        """
+        if run_dir == "":
+            # Use the synthesis directory for formal tools
+            run_dir = os.path.join(self.obj_dir, "syn-rundir")
```
Comment (colinschmidt) [inline src/hammer-vlsi/hammer_driver.py]: Why not make a new directory?

File: src/hammer-vlsi/hammer_driver.py [hunk 5]
```diff
@@ -148,6 +153,80 @@ def update_tool_configs(self) -> None:
...
+            tool_name=formal_tool_name
+        )
+        assert isinstance(formal_tool_get, HammerFormalTool), "Formal tool must be a HammerFormalTool"
+        formal_tool = formal_tool_get  # type: HammerTool
+        formal_tool.name = formal_tool_name
+        formal_tool.logger = self.log.context("formal")
+        formal_tool.set_database(self.database)
+        formal_tool.run_dir = run_dir
+
+        formal_tool.input_files = self.database.get_setting("synthesis.inputs.input_files")
```
Comment (colinschmidt) [inline src/hammer-vlsi/hammer_driver.py]: I don't think this overloading makes sense? But I'm not sure we need another option. Shouldn't the reference be the synthesis input and the implementation be the synthesis output?

File: src/hammer-vlsi/hammer_driver.py [hunk 6]
```diff
@@ -148,6 +153,80 @@ def update_tool_configs(self) -> None:
...
+
+        self.formal_tool = formal_tool
+
+        self.tool_configs["formal"] = formal_tool.get_config()
+        self.update_tool_configs()
+        return True
+
+    def load_power_tool(self, run_dir: str = "") -> bool:
+        """
+        Load the place and route tool based on the given database.
```
Comment (colinschmidt) [inline src/hammer-vlsi/hammer_driver.py]: `place and route` => `power analysis`

File: src/hammer-vlsi/hammer_driver.py [hunk 7]
```diff
@@ -148,6 +153,80 @@ def update_tool_configs(self) -> None:
...
+        )
+        assert isinstance(power_tool_get, HammerPowerTool), "Power tool must be a HammerPowerTool"
+        power_tool = power_tool_get  # type: HammerPowerTool
+        power_tool.name = power_tool_name
+        power_tool.logger = self.log.context("power")
+        power_tool.technology = self.tech
+        power_tool.set_database(self.database)
+        power_tool.run_dir = run_dir
+
+        # TODO: better way to get input files?
```
Comment (colinschmidt) [inline src/hammer-vlsi/hammer_driver.py]: I'll leave @edwardcwang to comment on a better way to do this. I assume its a similar parallel to `par.inputs`?

**Comment by donggyukim** [comment]
Sure, I'll add an example to the aspire repo. I'm out of country for visa renewal, so updates will be very slow though.

File: src/hammer-vlsi/hammer_driver.py (hunk 4 above)
Comment (donggyukim) [reply src/hammer-vlsi/hammer_driver.py]: The formality script is packaged in the DC script tarball.

File: src/hammer-vlsi/hammer_vlsi_impl.py [hunk 8]
```diff
@@ -1776,6 +1776,70 @@ def output_ilms(self, value: List[ILMStruct]) -> None:
             raise TypeError("output_ilms must be a List[ILMStruct]")
         self.attr_setter("_output_ilms", value)
 
+class HammerFormalTool(HammerTool):
```
[the remark this answers is not in the document]
Comment (edwardcwang) [reply src/hammer-vlsi/hammer_vlsi_impl.py]: @colinschmidt Do you think we can condense to just `HammerLECTool`?

**Comment by donggyukim** [comment]
It's such a pain. How can I fix it?

**Comment by donggyukim** [comment]
@hammer-robot Yes I win!

**Pull request closed without merging**